### PR TITLE
[JENKINS-72028] deprecated URL instantiation applied

### PR DESCRIPTION
**Pull request by abhishekmaity**
See [JENKINS-72028](https://issues.jenkins.io/browse/JENKINS-72028).

### Proposed changelog entries

- Entry 1: Issue, human-readable text

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/8515"></a>

### CODE REVIEW

File: core/src/test/java/jenkins/org/apache/commons/validator/routines/DomainValidatorTest.java [hunk 1]
```diff
@@ -25,9 +25,7 @@
 import java.io.InputStream;
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
-import java.net.HttpURLConnection;
-import java.net.IDN;
-import java.net.URL;
+import java.net.*;
```
Comment (NotMyFault) [inline core/src/test/java/jenkins/org/apache/commons/validator/routines/DomainValidatorTest.java:28]: That syntax is banned by checkstyle. How did you compile the PR successfully?

Comment (abhishekmaity) [reply core/src/test/java/jenkins/org/apache/commons/validator/routines/DomainValidatorTest.java]: @NotMyFault I have fixed the issue. Please re-review it. I compiled the PR as per instructions mentioned at [contribution guidelines](https://github.com/abhishekmaity/jenkins/blob/task-1/CONTRIBUTING.md) using openJDK 17 using IntelliJ IDEA

Comment (NotMyFault) [reply core/src/test/java/jenkins/org/apache/commons/validator/routines/DomainValidatorTest.java]: > I compiled the PR as per instructions mentioned at [contribution guidelines](https://github.com/abhishekmaity/jenkins/blob/task-1/CONTRIBUTING.md?rgh-link-date=2023-09-22T01%3A04%3A48Z) using openJDK 17 using IntelliJ IDEA

Jenkins does not compile if you invoke verification and there are code style issues.

**Comment by abhishekmaity** [comment]
Requesting @NotMyFault @Vlatombe or anyone from the @jenkinsci/core-pr-reviewers for a review

**Comment by abhishekmaity** [comment]
Request @NotMyFault @Vlatombe @MarkEWaite to close the PR

File: core/src/main/java/hudson/Functions.java [hunk 2]
```diff
@@ -1939,12 +1938,14 @@ public String getServerName() {
...
             if (url != null) {
-                String host = new URL(url).getHost();
+                String host = new URI(url).toURL().getHost();
                 if (host != null)
                     return host;
             }
         } catch (MalformedURLException e) {
             // fall back to HTTP request
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
Comment (timja) [inline core/src/main/java/hudson/Functions.java:1948]: should this also ignore the exception line `MalformedURLException` does?

**Review by timja — COMMENTED** [review]
looks ok, one question though

File: core/src/main/java/hudson/Functions.java (hunk 2 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/Functions.java:1948]: @timja, in this case error is wrapped in `RuntimeException` only because the result should invariably be a valid URL. So, addition of a checked exception type that would do nothing more than make the code more verbose as it would never be thrown. Therefore, it is opted to throw a `RuntimeException` instead.

Comment (timja) [reply core/src/main/java/hudson/Functions.java:1948]: but the `MalformedURLException` falls back to other code? i.e. this block is expected to be forgiving on invalid urls, e.g. empty strings.

Comment (abhishekmaity) [reply core/src/main/java/hudson/Functions.java:1948]: @timja please suggest if any on how to check that without breaking the code.

Comment (abhishekmaity) [reply core/src/main/java/hudson/Functions.java:1948]: @timja the block of code mentioned above is already marked `@Deprecated`. I made changes only related in resolution of [JENKINS-72028] issue.

Comment (timja) [reply core/src/main/java/hudson/Functions.java:1948]: I see, I couldn't find any usages of it so should be safe enough

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
Looks good so far, thanks for taking on the issue :)

File: core/src/main/java/hudson/PluginManager.java [hunk 3]
```diff
@@ -1955,6 +1955,8 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
                 }
             } catch (MalformedURLException e) {
                 return FormValidation.error(e.getMessage());
+            } catch (URISyntaxException e) {
+                throw new RuntimeException(e);
```
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1959]: why are we blowing up here and not disoplaying the error to the user?

**Review by jtnord — DISMISSED** [review]

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 4]
```diff
@@ -135,9 +136,11 @@ public String getAdvertisedHost() {
           return CLI_HOST_NAME;
         }
         try {
-            return new URL(Jenkins.get().getRootUrl()).getHost();
+            return new URI(Jenkins.get().getRootUrl()).toURL().getHost();
         } catch (MalformedURLException e) {
             throw new IllegalStateException("Could not get TcpSlaveAgentListener host name", e);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java:143]: does not help diagnose an ussue? - add a message like above?
```suggestion
             throw new IllegalStateException("Could not get TcpSlaveAgentListener host name", e);
```

File: core/src/main/java/jenkins/security/ResourceDomainConfiguration.java [hunk 5]
```diff
@@ -112,9 +114,11 @@ private FormValidation checkUrl(String resourceRootUrlString, boolean allowOnlin
 
         URL resourceRootUrl;
         try {
-            resourceRootUrl = new URL(resourceRootUrlString);
+            resourceRootUrl = new URI(resourceRootUrlString).toURL();
         } catch (MalformedURLException ex) {
             return FormValidation.error(Messages.ResourceDomainConfiguration_Invalid());
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
Comment (jtnord) [inline core/src/main/java/jenkins/security/ResourceDomainConfiguration.java:121]: this is wrong and regresses functionality - the error needs to be user visible so they can fix it.

**Comment by jtnord** [comment]
I found several of the same types of issues and so did not fully review all the code, so the list of comments about throwing an exception where it is inappropriate to throw one should be considered potentially incomplete.

In general - something that returns a `FormValidation` should return a `FormValidation` and never throw an exception.
Methods that previously throw a `MalformedUrlException` should probably still throw that rather than a `RuntimeException` - as `MalformedUrlException` is a checked exception callers must already be handling the case that the url they provide is garbage and taking appropriate actions.

File: core/src/main/java/hudson/PluginManager.java (hunk 3 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/PluginManager.java]: I have replaced
```
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
```
with
```
        } catch (URISyntaxException e) {
            throw new URISyntaxException(e.getReason(), e.getInput());
```

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 4 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java]: I applied 
            `throw new URISyntaxException(e.getReason(), e.getInput());`

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 6]
```diff
@@ -2422,7 +2428,11 @@ public PluginDowngradeJob(Plugin plugin, UpdateSite site, Authentication auth) {
 
         @Override
         protected URL getURL() throws MalformedURLException {
-            return new URL(plugin.url);
+            try {
+                return new URI(plugin.url).toURL();
+            } catch (URISyntaxException e) {
+                throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (abhishekmaity) [reply core/src/main/java/hudson/model/UpdateCenter.java:2434]: I have included both `MalformedURLException` and `URISyntaxException` together in catch followed by throwing `RuntimeException`.

[the remark this answers is not in the document]
Comment (abhishekmaity) [reply core/src/main/java/hudson/model/UpdateCenter.java:2434]: @jtnord we cannot only use `MalformedURLException` here because it causes un-handled error `URISyntaxException` arising from usage of `.toURL()`

File: core/src/main/java/jenkins/security/ResourceDomainConfiguration.java (hunk 5 above)
Comment (abhishekmaity) [reply core/src/main/java/jenkins/security/ResourceDomainConfiguration.java]: @jtnord  I have replaced
```
        } catch (URISyntaxException e) {
            throw new RuntimeException(e);
```
with
```
        } catch (URISyntaxException e) {
            throw new URISyntaxException(e.getReason(), e.getInput());
```
Thus, used visibility is covered now for the specific exception i.e. `URISyntaxException`

**Comment by abhishekmaity** [comment]
> I found several of the same types of issues and so did not fully review all the code, so the list of comments about throwing an exception where it is inappropriate to throw one should be considered potentially incomplete.
> 
> In general - something that returns a `FormValidation` should return a `FormValidation` and never throw an exception. Methods that previously throw a `MalformedUrlException` should probably still throw that rather than a `RuntimeException` - as `MalformedUrlException` is a checked exception callers must already be handling the case that the url they provide is garbage and taking appropriate actions.

I tried to replace generic `RuntimeException` with the specific exception at all places where `.toURL` is applied, but then it is causing a domino effect of changes/errors in other java files due to dependency.  Everytime I fix, one error pop up in from other files after compilation.

**Comment by abhishekmaity** [comment]
@NotMyFault @timja Please merge the PR

**Comment by abhishekmaity** [comment]
@jtnord any feedback on the above changes made.

File: cli/src/main/java/hudson/cli/SSHCLI.java [hunk 7]
```diff
@@ -86,12 +90,9 @@ static int sshConnection(String jenkinsUrl, String user, List<String> args, Priv
...
 
-            KnownHostsServerKeyVerifier verifier = new DefaultKnownHostsServerKeyVerifier(new ServerKeyVerifier() {
-                @Override
-                public boolean verifyServerKey(ClientSession clientSession, SocketAddress remoteAddress, PublicKey serverKey) {
-                    CLI.LOGGER.log(Level.WARNING, "Unknown host key for {0}", remoteAddress.toString());
-                    return !strictHostKey;
-                }
+            KnownHostsServerKeyVerifier verifier = new DefaultKnownHostsServerKeyVerifier((clientSession, remoteAddress, serverKey) -> {
+                CLI.LOGGER.log(Level.WARNING, "Unknown host key for {0}", remoteAddress.toString());
+                return !strictHostKey;
```
Comment (jtnord) [inline cli/src/main/java/hudson/cli/SSHCLI.java:95]: unrelated change?

File: core/src/main/java/hudson/FilePath.java [hunk 8]
```diff
@@ -1065,6 +1066,8 @@ private boolean installIfNecessaryFrom(@NonNull URL archive, @NonNull TaskListen
             return true;
         } catch (IOException e) {
             throw new IOException("Failed to install " + archive + " to " + remote, e);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:1070]: you should never be throwing this.

File: core/src/main/java/hudson/Functions.java [hunk 9]
```diff
@@ -1933,18 +1932,21 @@ public static String toEmailSafeString(String projectName) {
      * @deprecated use {@link JNLPLauncher#getInboundAgentUrl}
      */
     @Deprecated
-    public String getServerName() {
+    public String getServerName() throws URISyntaxException, MalformedURLException {
```
Comment (jtnord) [inline core/src/main/java/hudson/Functions.java:1935]: this is a breaking API change which you can not do as the method is public and can be used outside the scope of this project.

File: core/src/main/java/hudson/Functions.java [hunk 10]
```diff
@@ -1933,18 +1932,21 @@ public static String toEmailSafeString(String projectName) {
...
         try {
             if (url != null) {
-                String host = new URL(url).getHost();
+                String host = new URI(url).toURL().getHost();
                 if (host != null)
                     return host;
             }
         } catch (MalformedURLException e) {
             // fall back to HTTP request
+            throw new MalformedURLException();
```
Comment (jtnord) [inline core/src/main/java/hudson/Functions.java:1947]: this is not falling back.

File: core/src/main/java/hudson/Functions.java [hunk 11]
```diff
@@ -1933,18 +1932,21 @@ public static String toEmailSafeString(String projectName) {
...
+        } catch (URISyntaxException e) {
+            throw new URISyntaxException(e.getReason(), e.getInput());
```
Comment (jtnord) [inline core/src/main/java/hudson/Functions.java:1949]: you need to handle this and not throw it.  

fallback to the current server name as we do today.

File: core/src/main/java/hudson/PluginManager.java [hunk 12]
```diff
@@ -1955,6 +1955,8 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
                 }
             } catch (MalformedURLException e) {
                 return FormValidation.error(e.getMessage());
+            } catch (URISyntaxException e) {
+                throw new URISyntaxException(e.getReason(), e.getInput());
```
Comment (jtnord) [inline core/src/main/java/hudson/PluginManager.java:1959]: you need to handle this as appropriate for the function - do not rethrow this.

in other words the handling of this exception is returning a `FormValidation` error with the message so that the user can understand they have entered giberish into the form

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 13]
```diff
@@ -130,14 +131,16 @@ public int getAdvertisedPort() {
      * Gets the host name that we advertise protocol clients to connect to.
      * @since 2.198
      */
-    public String getAdvertisedHost() {
+    public String getAdvertisedHost() throws URISyntaxException {
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java:134]: this is a breaking change.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java [hunk 14]
```diff
@@ -130,14 +131,16 @@ public int getAdvertisedPort() {
...
         if (CLI_HOST_NAME != null) {
           return CLI_HOST_NAME;
         }
         try {
-            return new URL(Jenkins.get().getRootUrl()).getHost();
+            return new URI(Jenkins.get().getRootUrl()).toURL().getHost();
         } catch (MalformedURLException e) {
             throw new IllegalStateException("Could not get TcpSlaveAgentListener host name", e);
+        } catch (URISyntaxException e) {
```
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java:142]: this should be handled the same as if the URL provided was garbage - see the line above,  you can even use multi-catch here

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 15]
```diff
@@ -2193,7 +2195,11 @@ public InstallationJob(Plugin plugin, UpdateSite site, Authentication auth, bool
...
```
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateCenter.java:2201]: callers of this already had to cope and adjust to the URL being invalid - they can not longer do this and their code will be broken as this will thow a RuntimeException and no longer MalformedURLException.

rather than throwing the RuntimeExeption throw a `MalformedURLException` with the appropriate message initialized from this exception

File: core/src/main/java/hudson/model/UpdateCenter.java [hunk 16]
```diff
@@ -2422,7 +2428,11 @@ public PluginDowngradeJob(Plugin plugin, UpdateSite site, Authentication auth) {
 
         @Override
         protected URL getURL() throws MalformedURLException {
-            return new URL(plugin.url);
+            try {
+                return new URI(plugin.url).toURL();
+            } catch (MalformedURLException | URISyntaxException e) {
+                throw new RuntimeException(e);
```
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateCenter.java:2434]: same as above - you should never be throwing RuntimeExeption where the method was previously docuemnted to throw a MalformedURLException

**Review by jtnord — CHANGES_REQUESTED** [review]

File: cli/src/main/java/hudson/cli/SSHCLI.java [hunk 17]
```diff
@@ -60,9 +59,14 @@
...
 class SSHCLI {
 
-    static int sshConnection(String jenkinsUrl, String user, List<String> args, PrivateKeyProvider provider, final boolean strictHostKey) throws IOException {
+    static int sshConnection(String jenkinsUrl, String user, List<String> args, PrivateKeyProvider provider, final boolean strictHostKey) throws IOException, URISyntaxException {
         Logger.getLogger(SecurityUtils.class.getName()).setLevel(Level.WARNING); // suppress: BouncyCastle not registered, using the default JCE provider
-        URL url = new URL(jenkinsUrl + "login");
+        URL url = null;
+        try {
+            url = new URI(jenkinsUrl + "login").toURL();
+        } catch (URISyntaxException e) {
```
[the remark this answers is not in the document]
Comment (abhishekmaity) [reply cli/src/main/java/hudson/cli/SSHCLI.java:65]: I have made the below changes to above code now per suggestion
```
        URL url = null;
        try {
            url = new URI(jenkinsUrl + "login").toURL();
        } catch (URISyntaxException e) {
            throw new URISyntaxException(e.getInput(),e.getReason());
```

File: cli/src/main/java/hudson/cli/SSHCLI.java (hunk 7 above)
Comment (abhishekmaity) [reply cli/src/main/java/hudson/cli/SSHCLI.java]: Reverted to original code now. Tried to implement unecessary lambda for the function on automated suggestion.

File: core/src/main/java/hudson/FilePath.java (hunk 8 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/FilePath.java]: Instead now throwing the following
`throw new URISyntaxException(e.getInput(), e.getReason());`

File: core/src/main/java/hudson/Functions.java (hunk 9 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/Functions.java]: Removed `throws URISyntaxException, MalformedURLException `

File: core/src/main/java/hudson/Functions.java (hunk 10 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/Functions.java]: removed `throw new MalformedURLException();` to enable fallback as in original code

File: core/src/main/java/hudson/Functions.java (hunk 11 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/Functions.java]: Replaced with the following
```
} catch (MalformedURLException | URISyntaxException e) {
            // fall back to HTTP request
        }
```

File: core/src/main/java/hudson/PluginManager.java (hunk 12 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/PluginManager.java]: Replaced with below as explained above
```
} catch (MalformedURLException |URISyntaxException e) {
                return FormValidation.error(e.getMessage());
            }
```

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 13 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java]: Removed  `throws URISyntaxException` and modified the block of code as below
```
public String getAdvertisedHost() {
        if (CLI_HOST_NAME != null) {
          return CLI_HOST_NAME;
        }
        try {
            return new URI(Jenkins.get().getRootUrl()).toURL().getHost();
        } catch (MalformedURLException | URISyntaxException e) {
            throw new IllegalStateException("Could not get TcpSlaveAgentListener host name", e);
        }
    }
```

File: core/src/main/java/hudson/TcpSlaveAgentListener.java (hunk 14 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java]: Done. See below
```
} catch (MalformedURLException | URISyntaxException e) {
            throw new IllegalStateException("Could not get TcpSlaveAgentListener host name", e);
        }
```

File: core/src/main/java/hudson/model/UpdateCenter.java (hunk 15 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/model/UpdateCenter.java]: Changed to
```
            try {
                return new URI(plugin.url).toURL();
            } catch (URISyntaxException e) {
                throw new MalformedURLException("Not a valid URL");
            }
```

File: core/src/main/java/hudson/model/UpdateCenter.java (hunk 16 above)
Comment (abhishekmaity) [reply core/src/main/java/hudson/model/UpdateCenter.java]: Changed to
```
            try {
                return new URI(plugin.url).toURL();
            } catch (URISyntaxException e) {
                throw new MalformedURLException("Not a valid URL");
            }

**Comment by abhishekmaity** [comment]
@jtnord I have made changes based on the review. Re-request to have a look and ready PR for merge.

**Review by jtnord — CHANGES_REQUESTED** [review]
There are many similar issues throughout the PR, I have been commenting on occurrences of the types of issue but not all the issues but this has seemingly caused confusion thinking they where only the issues.

So.

--------------

unless you are explicitly redacting/hiding information (which is not the case in this PR)


Do not write code like:

```java
try {
  // something
catch (SomeException e) {
  throw new SomeException(....)
}
```
as it makes the code worse - especially when the code inside the `try` block is long and the exception could have been thrown on any one of a number of lines.

rather just keep the code as

```java
// something
```

--------------

non internal code (ie anything protected or public) can not have its method signature changed.  This includes adding a new `Exception` type to the declared `throws` that is not a subclass of an  existing declared one.

e.g. if the method was

```java
public void doSomething(String url) throws MalformedURLException {
 // something
 }
```

then you can not change it to 

```java
public void doSomething(String url) throws MalformedURLException, URISyntaxException {
 // something
 }
 ```

in this case and so that callers can react the same way to an "invalid URL"  you should probably change the code so that it looks like

```java
public void doSomething(String url) throws MalformedURLException {
  try {
    // something
  } catch (URISyntaxException e) {
    MalformedURLException mex = new MalformedURLException(e.getMessage());
    mex.initCause(e);
    throw mex;
  }
 }
 ```

-----

Internal code can have its method changed without breaking plugins.

**However** if you change the signature to throw a `URISyntaxException`  you need to ensure that **every** caller that calls the method handles a `URISyntaxException` in **exactly** the same way as if it had caught a `MalformedURLException` (if it has special handling for this);

As `MalformedURLException` is an `IOException` this may not be obvious.

You could handle the `URISyntaxException` differently, but this would need a case by case analysis and would generally only be applicable if the code was catching an `IOException`, but even in this case the code needs to take _appropriate_ action.

**Comment by jtnord** [comment]
dismissing previous approvaals as the code has moved on so much.  (not asking for a review at this stage as it is not ready)

**Comment by abhishekmaity** [comment]
@jtnord @NotMyFault @timja I am closing this PR and will open a new PR a fresh based on the observation. Many thanks @jtnord for taking out time and highlighting the issues. Instead of a single PR, I am thinking to do smaller PR.
